internal/controller/deployment_ to_pdb_controller.go

```diff
@@ -0,0 +1,166 @@
+		Name:      deployment.Name + "-pdb",
+	}, pdb)
+
+	if err == nil {
+		// PDB already exists, nothing to do
+		log.Info("PodDisruptionBudget already exists", "namespace", deployment.Namespace, "name", deployment.Name)
+		//if pdb.Spec.MinAvailable.IntVal != *deployment.Spec.Replicas {
+		//	pdb.Spec.MinAvailable.IntVal = *deployment.Spec.Replicas
+		//	if err := r.Update(ctx, pdb); err != nil {
+		//		return reconcile.Result{}, err
+		//	}
+		//}
+		return reconcile.Result{}, nil
+	}
+
+	// Create a new PDB for the Deployment
+	pdb = &policyv1.PodDisruptionBudget{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "PodDisruptionBudget",
+			APIVersion: "policy/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      r.generatePDBName(deployment.Name),
+			Namespace: deployment.Namespace,
+			Annotations: map[string]string{
+				"createdBy": "DeploymentToPDBController",
+				"target":    deployment.Name,
+			},
+		},
+		Spec: policyv1.PodDisruptionBudgetSpec{
+			MinAvailable: &intstr.IntOrString{IntVal: *deployment.Spec.Replicas},
+			Selector:     &metav1.LabelSelector{MatchLabels: deployment.Spec.Selector.MatchLabels},
+		},
+	}
+
+	log.Info("Creating PodDisruptionBudget", "namespace", pdb.Namespace, "name", pdb.Name)
+	if err := r.Create(ctx, pdb); err != nil {
+		log.Error(err, "unable to create PDB")
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (r *DeploymentToPDBReconciler) generatePDBName(deploymentName string) string {
+	return deploymentName + "-pdb"
+}
+
+// handleDeploymentDeletion deletes the associated PDB when the Deployment is deleted
+func (r *DeploymentToPDBReconciler) handleDeploymentDeletion(ctx context.Context, req ctrl.Request) (reconcile.Result, error) {
+	log := log.FromContext(ctx)
+
+	// Try to get the associated PDB
+	pdb := &policyv1.PodDisruptionBudget{}
+	err := r.Get(ctx, client.ObjectKey{
+		Namespace: req.NamespacedName.Namespace,
+		Name:      r.generatePDBName(req.NamespacedName.Name),
+	}, pdb)
+
+	if err != nil {
+		// If there's no PDB or it can't be fetched, just return
+		log.Info("PodDisruptionBudget does not exist or error fetching", "namespace", req.NamespacedName.Namespace, "name", req.NamespacedName.Name)
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// If the PDB exists, delete it
+	log.Info("Deleting PodDisruptionBudget", "namespace", pdb.Namespace, "name", pdb.Name)
+	if err := r.Delete(ctx, pdb); err != nil {
+		log.Error(err, "unable to delete PDB")
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *DeploymentToPDBReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	logger := mgr.GetLogger()
+	// Set up the controller to watch Deployments and trigger the reconcile function
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1.Deployment{}).
+		WithEventFilter(predicate.Funcs{
+			// Only trigger for Create and Delete events
```

Comment:
i THINK the default is to handle create/deletes so you don't need the creatfunc/deletefunc unless you want this logging? 

Also note that I think when your controller restarts everything is seen as a crete event? (I'm not actually sure if this is true or not)

Reply:
Still applies can just remove everything but updatefunc?